homeassistant/components/doorbird/event.py

```diff
@@ -0,0 +1,88 @@
+"""Support for doorbird events."""
+
+from typing import TYPE_CHECKING
+
+from homeassistant.components.event import (
+    EventDeviceClass,
+    EventEntity,
+    EventEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .device import DoorbirdEvent
+from .entity import DoorBirdEntity
+from .models import DoorBirdData
+
+EVENT_DESCRIPTIONS = {
+    "doorbell": EventEntityDescription(
+        key="doorbell",
+        translation_key="doorbell",
+        device_class=EventDeviceClass.DOORBELL,
+        event_types=["ring"],
+    ),
+    "motion": EventEntityDescription(
+        key="motion",
+        translation_key="motion",
+        device_class=EventDeviceClass.MOTION,
+        event_types=["motion"],
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the DoorBird event platform."""
+    config_entry_id = config_entry.entry_id
+    door_bird_data: DoorBirdData = hass.data[DOMAIN][config_entry_id]
+    async_add_entities(
+        DoorBirdEventEntity(door_bird_data, doorbird_event, description)
+        for doorbird_event in door_bird_data.door_station.event_descriptions
+        if (description := EVENT_DESCRIPTIONS.get(doorbird_event.event_type))
+    )
+
+
+class DoorBirdEventEntity(DoorBirdEntity, EventEntity):
+    """A relay in a DoorBird device."""
+
+    entity_description: EventEntityDescription
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        door_bird_data: DoorBirdData,
+        doorbird_event: DoorbirdEvent,
+        entity_description: EventEntityDescription,
+    ) -> None:
+        """Initialize an event for a doorbird device."""
+        super().__init__(door_bird_data)
+        self._doorbird_event = doorbird_event
+        self.entity_description = entity_description
+        event = doorbird_event.event
+        self._attr_unique_id = f"{self._mac_addr}_{event}"
+        slug_name = event.removeprefix(self._door_station.slug).strip("_")
+        friendly_name = slug_name.replace("_", " ")
+        self._attr_name = friendly_name[0:1].upper() + friendly_name[1:].lower()
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to device events."""
+        self.async_on_remove(
+            self.hass.bus.async_listen(
+                f"{DOMAIN}_{self._doorbird_event.event}",
```

Comment:
We shouldn't use the event bus for internal communication in the integration.

Reply:
Future PR is planned to get rid of the custom events. I haven’t come up with a good replacement for the camera urls or logbook yet.

Reply:
Camera urls should be easy to move over.  The event entities and logbook need some more work with logbook in general as there is no way to describe them currently. That’s not a DoorBird specific issue though.

Everything in this integration has historically used events to communicate…. It’s quite old…I could replace all the internal communication with the dispatcher in the mean time but it seemed a bit much do fire an event and also use the dispatcher.

Reply:
Its going to take longer than I had hoped to figure out the logbook so I'm going to open a PR to add the dispatcher. We will fire the event and dispatch until I can come up with a way to get rid of the custom events. We might have to bit the bullet on that one and do a breaking change to remove them in a few months after the event entities have been around for a while if they work out as planned

Reply:
https://github.com/home-assistant/core/pull/121825